server/schedulers/transfer_witness_leader.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule"
+	"github.com/tikv/pd/server/schedule/filter"
+	"github.com/tikv/pd/server/schedule/operator"
+	"github.com/tikv/pd/server/schedule/plan"
+	"github.com/tikv/pd/server/storage/endpoint"
+)
+
+const (
+	// TransferWitnessLeaderName is transfer witness leader scheduler name.
+	TransferWitnessLeaderName = "transfer-witness-leader-scheduler"
+	// TransferWitnessLeaderType is transfer witness leader scheduler type.
+	TransferWitnessLeaderType = "transfer-witness-leader"
+	// TransferWitnessLeaderBatchSize is the number of operators to to transfer
+	// leaders by one scheduling
+	TransferWitnessLeaderBatchSize = 3
+	// TransferWitnessLeaderRecvMaxRegionSize is the max number of region can receive
+	// TODO: make it a reasonable value
+	TransferWitnessLeaderRecvMaxRegionSize = 1000
+)
+
+func init() {
+	schedule.RegisterSliceDecoderBuilder(TransferWitnessLeaderType, func(args []string) schedule.ConfigDecoder {
+		return func(v interface{}) error {
+			return nil
+		}
+	})
+
+	schedule.RegisterScheduler(TransferWitnessLeaderType, func(opController *schedule.OperatorController, _ endpoint.ConfigStorage, _ schedule.ConfigDecoder) (schedule.Scheduler, error) {
+		return newTransferWitnessLeaderScheduler(opController), nil
+	})
+}
+
+type transferLeaderScheduler struct {
+	*BaseScheduler
+	regions chan *core.RegionInfo
+}
+
+// newTransferWitnessLeaderScheduler creates an admin scheduler that transfers witness leader of a region.
+func newTransferWitnessLeaderScheduler(opController *schedule.OperatorController) schedule.Scheduler {
+	return &transferLeaderScheduler{
+		BaseScheduler: NewBaseScheduler(opController),
+		regions:       make(chan *core.RegionInfo, TransferWitnessLeaderRecvMaxRegionSize),
+	}
+}
+
+func (s *transferLeaderScheduler) GetName() string {
+	return TransferWitnessLeaderName
+}
+
+func (s *transferLeaderScheduler) GetType() string {
+	return TransferWitnessLeaderType
+}
+
+func (s *transferLeaderScheduler) IsScheduleAllowed(cluster schedule.Cluster) bool {
+	// TODO: make sure the restriction is reasonable
+	allowed := s.OpController.OperatorCount(operator.OpLeader) < cluster.GetOpts().GetLeaderScheduleLimit()
```

Comment:
It could be blocked if there are balance leader operators.

Reply:
Yes, wait for the scene test to observe, and then determine how to adjust, add a TODO